client/cos.go

```diff
@@ -0,0 +1,261 @@
+// Copyright 2024 OpenPubkey
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package client
+
+import (
+	"context"
+	"crypto"
+	"crypto/rand"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/lestrrat-go/jwx/v2/jws"
+	"github.com/openpubkey/openpubkey/cosigner/msgs"
+	"github.com/openpubkey/openpubkey/pktoken"
+	"github.com/openpubkey/openpubkey/util"
+	"github.com/sirupsen/logrus"
+)
+
+type CosignerProvider struct {
+	Issuer       string
+	CallbackPath string
+}
+
+func (p *CosignerProvider) GetIssuer() string {
+	return p.Issuer
+}
+
+func (c *CosignerProvider) RequestToken(ctx context.Context, signer crypto.Signer, pkt *pktoken.PKToken, redirCh chan string) (*pktoken.PKToken, error) {
+	// Find an unused port
+	listener, err := net.Listen("tcp", ":0")
+	if err != nil {
+		return nil, fmt.Errorf("failed to bind to an available port: %w", err)
+	}
```

Comment:
I think anything with a server in it likely belongs in the examples folder. Our core code should be proposing a framework with specific parts filled out and others left blank for customers to plug into. I think that was a big part of the V1 architecture. Here, I think we're assuming too much and aren't building in the flexibility this kind of library should have.

Reply:
This is following the HTTP server pattern we already use in the client to get the callback from the OP. 

```golang
	logrus.Infof("listening on http://%s/", lis)
	logrus.Info("press ctrl+c to stop")
	earl := fmt.Sprintf("http://localhost:%s/login", g.RedirURIPort)
	util.OpenUrl(earl)

	go func() {
		err := g.server.ListenAndServe()
		if err != nil && err != http.ErrServerClosed {
			logrus.Error(err)
		}
	}()
```
https://github.com/openpubkey/openpubkey/blob/main/client/providers/google.go#L92

Reply:
Yup, I also think that should be in examples but I was just making a comment on the large amount of new http code in this PR.